### PR TITLE
feat(pubsub): Enable PubSub to work along side GraphQL/Appsync

### LINKED ISSUE #823: Amplify PubSub not compatible with Amplify API GraphQL / Appsync
Feature Request - enable PubSub to work along side GraphQL/Appsync

Having a REST API, I added PubSub and observed the expected PubSub functionality when unauthenticated, and after adding both IAM and IOT policies and attaching the Cognito ID of each subscriber, I also observed expected PubSub functionality when authenticated.

Having upgraded from REST to AppSync/GraphQL, I observe the desired Query and Mutation behaviour. Despite many subscription/authorization patterns available, as far as I can tell the GraphQL subscription mechanism will not be flexible enough to meet my authorization needs. 

In order to meet my needs, I returned to the more generic PubSub. But I have discovered that under the same PubSub configuration, the PubSub functionality that worked along side REST no longer works along side GraphQL. I suspect that Amplify PubSub and Amplify API GraphQL have contention in using the same MQTT/WebSocket resources.

I have the correct PubSub configuration:
```
Amplify.addPluggable(
  new AWSIoTProvider({
    aws_pubsub_region: <my region>
    aws_pubsub_endpoint: <my endpoint>
  })
);
```
With GraphQL (but not with REST) subscribing to a PubSub Topic results in this error:
Error: Missing connections info
    at AWSAppSyncProvider../node_modules/aws-amplify/lib/PubSub/Providers/AWSAppSyncProvider.js.AWSAppSyncProvider.subscribe (AWSAppSyncProvider.js:75)

### LINKED ISSUE #1404: Pubsub: {} in Amplify.configure conflict with appsync subscriptions
This is _partially_ related to #823

Do you want to request a feature or report a bug?
**Bug**

What is the current behavior?

This configuration throws an error, while the configuration below it works fine
```
Amplify.configure({
  Auth: {
    identityPoolId: 
    region: 
    userPoolId: 
    userPoolWebClientId: 
  },
  PubSub: {
    aws_pubsub_region: 
    aws_pubsub_endpoint: 
 },

  aws_appsync_graphqlEndpoint:
  aws_appsync_region: 
  aws_appsync_authenticationType: 
  aws_pubsub_endpoint: 
});
```

```
Amplify.configure({
  Auth: {
    identityPoolId: 
    region: 
    userPoolId: 
    userPoolWebClientId: 
  },
//  PubSub: {
//    aws_pubsub_region: 
//    aws_pubsub_endpoint: 
// },

  aws_appsync_graphqlEndpoint:
  aws_appsync_region: 
  aws_appsync_authenticationType: 
  aws_pubsub_endpoint: 
});

```
**With the second configuration, I can subscribe to Appsync graphql endpoint**

If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than AWS Amplify.

```

let subquery = `a subscription`
Auth.signIn("username", "password").then(data =>{
  try {
      let sub = API.graphql(graphqlOperation(subquery));
      sub.subscribe({
        next: data => {
          console.log(data);
        },
        error: err => console.log(err)
      });
    } catch (e) {
      console.log(e);
    }
}

```

What is the expected behavior?

Subscriptions should work normally

Which versions of Amplify, and which browser / OS are affected by this issue? Did this work in previous versions?

Tested on ReactNative

### LINKED ISSUE #2124: AppSync Subscriptions not working when Rest API and AppSync are set simultaneously
####  Do you want to request a feature or report a bug?
Bug

####   What is the current behavior?
AppSync Subscriptions not working when Rest API and AppSync  are set simultaneously using Amplify.configure. Rest API, mutation and query are working below setting. But subscription is not work.

```
Amplify.configure({
  API: {
    endpoints: [
      {
        name: 'example',
        endpoint: 'http://localhost:5000',
        custom_header: async () => {
          const token = (await Auth.currentSession())['idToken'].jwtToken;
          return {
            Authorization: `Bearer ${token}`,
          };
        },
      },
    ],
    aws_appsync_apiKey: <REACT_APP_AWS_APPSYNC_APIKEY>,
    aws_appsync_authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    aws_appsync_graphqlEndpoint: <REACT_APP_AWS_APPSYNC_GRAPHQLENDPOINT>,
    aws_appsync_region: <REACT_APP_AWS_APPSYNC_REGION>,
  },
  Auth: {
    identityPoolId: <REACT_APP_IDENTITY_POOL_ID>,
    region: <REACT_APP_AWS_REGION>,
    userPoolId: <REACT_APP_USER_POOL_ID>,
    userPoolWebClientId:<REACT_APP_USER_WEB_CLIENT_ID>,
  },
});
```

If I delete the Rest setting, subscription will work.
```
Amplify.configure({
  aws_appsync_apiKey: <REACT_APP_AWS_APPSYNC_APIKEY>,
  aws_appsync_authenticationType: 'AMAZON_COGNITO_USER_POOLS',
  aws_appsync_graphqlEndpoint: <REACT_APP_AWS_APPSYNC_GRAPHQLENDPOINT>,
  aws_appsync_region: <REACT_APP_AWS_APPSYNC_REGION>,
  Auth: {
    identityPoolId: <REACT_APP_IDENTITY_POOL_ID>,
    region: <REACT_APP_AWS_REGION>,
    userPoolId: <REACT_APP_USER_POOL_ID>,
    userPoolWebClientId:<REACT_APP_USER_WEB_CLIENT_ID>,
  },
});
```

If AppSync configuration is placed at the root of the configuration, AppSync endpoint will not be resolved.

```
Amplify.configure({
  API: {
    endpoints: [
      {
        name: 'example',
        endpoint: 'http://localhost:5000',
        custom_header: async () => {
          const token = (await Auth.currentSession())['idToken'].jwtToken;
          return {
            Authorization: `Bearer ${token}`,
          };
        },
      },
    ],
    aws_appsync_apiKey: <REACT_APP_AWS_APPSYNC_APIKEY>,
    aws_appsync_authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    aws_appsync_graphqlEndpoint: <REACT_APP_AWS_APPSYNC_GRAPHQLENDPOINT>,
    aws_appsync_region: <REACT_APP_AWS_APPSYNC_REGION>,
  },
  Auth: {
    identityPoolId: <REACT_APP_IDENTITY_POOL_ID>,
    region: <REACT_APP_AWS_REGION>,
    userPoolId: <REACT_APP_USER_POOL_ID>,
    userPoolWebClientId:<REACT_APP_USER_WEB_CLIENT_ID>,
  },
});
```

I referred to https://github.com/aws-amplify/amplify-js/issues/2012 , but the problem was not solved.

### PR DESCRIPTION
**Pull request by alebiavati**
*Issue #, if available:*
fix #823
fix #2124
fix #1404

*Description of changes:*

The issues above can be summarized with following points:
* The PubSub category allows to add multiple providers, but then when a developer subscribes/publishes to a topic, PubSub will subscribe/publish to ALL registered providers at the same time. There is no way currently to control which provider to use.
* The API category uses the PubSub category for subscribing to messages from AppSync, which means that an AppSync provider is added to the PubSub category during initialization. This makes it impossible for a developer to use another PubSub Provider for a separate connection via MQTT or AWS IoT.

With this PR developers can specify a provider name when calling PubSub subscribe/publish:

```js
Amplify.addPluggable(new AWSIotProvider({
  aws_pubsub_region: 'us-east-1',
  aws_pubsub_endpoint: 'wss://aws-iot-endpoint',
}));

Amplify.addPluggable(new MqttOverWSProvider({
  aws_pubsub_endpoint: 'wss://my-custom-mqtt',
}));

// Then they can publish/subscribe to each specific provider
PubSub.subscribe('my-channel', {
  provider: 'AWSIotProvider'
}).subscribe(...);

PubSub.subscribe('my-channel', {
  provider: 'MqttOverWSProvider'
}).subscribe(...);
```

If no `provider` option is given, PubSub will subscribe/publish to all available providers that were added by the developer.

Internally, the API category uses this feature for AppSync subscriptions. In particular, the API category uses an internal-only provider of type `AWSAppSyncProvider` by calling:

```js
PubSub.subscribe('my-channel', {
  provider: INTERNAL_AWS_APPSYNC_PUBSUB_PROVIDER,
}).subscribe(...);
```

where `INTERNAL_AWS_APPSYNC_PUBSUB_PROVIDER` is an internal Symbol.

Note that with this PR I also change where the `AWSAppSyncProvider` is instantiated. Before we checked for the presence of AppSync configurations, but now we only check for the Symbol above and the `AWSAppSyncProvider` is instantiated lazily and cached after the first AppSync subscription.

### Breaking changes

If a developer adds the `AWSAppSyncProvider` in their app via `Amplify.addPluggable`, this provider will NOT be used by the API category to subscribe to AppSync. Instead another internal-only provider of type `AWSAppSyncProvider` is added to the PubSub category, then the API category will only use this internal provider.

Before:

```js
// app's index.js
Amplify.addPluggable(new AWSAppSyncProvider({
  // custom user configuration
}))

// The custom provider added above will be used to subscribe to AppSync.
export default () => <Connect query={...}>{...}</Connect>
```

After:
```js
// app's index.js
Amplify.addPluggable(new AWSAppSyncProvider({
  // custom user configuration
}))

// The custom provider added  above by the developer is ignored,
// instead an internal-only provider will be used to subscribe to AppSync.
export default () => <Connect query={...}>{...}</Connect>
```

I don't think this is a common use case, especially because the `AWSAppSyncProvider` doesn't actually accept any special configuration.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/pubsub/src/Providers/AWSAppSyncProvider.ts [hunk 1]
```diff
@@ -19,10 +19,11 @@ import { MqttOverWSProvider } from './MqttOverWSProvider';
 const logger = new Logger('AWSAppSyncProvider');
 
 export class AWSAppSyncProvider extends MqttOverWSProvider {
+    static PROVIDER_NAME = 'AWSAppSyncProvider';
 
     protected get endpoint() { throw new Error('Not supported'); }
 
-    getProviderName() { return 'AWSAppSyncProvider'; }
+    getProviderName() { return this.options.name || AWSAppSyncProvider.PROVIDER_NAME; }
```
Comment (alebiavati) [inline packages/pubsub/src/Providers/AWSAppSyncProvider.ts]: Question: would it be better to use a more specific option name like `aws_pubsub_name`?

File: packages/core/src/constants.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+/**
+ * This constant is used to create an internal-only PubSub provider that 
+ * is used for AppSync/GraphQL subscriptions in the API category.
+ */
+export const INTERNAL_AWS_APPSYNC_PUBSUB_PROVIDER = 'INTERNAL_AWS_APPSYNC_PUBSUB_PROVIDER';
```
Comment (alebiavati) [inline packages/core/src/constants.ts]: I couldn't find any other place where to share **internal** constants between multiple categories. This constant should NOT be available in the `Amplify` object as it is for internal use only.

**Comment by powerful23** [comment]
Docs at https://github.com/aws-amplify/docs/pull/644

File: packages/pubsub/src/PubSub.ts [hunk 3]
```diff
@@ -53,11 +69,6 @@ export default class PubSub {
 
         this._options = Object.assign({}, this._options, opt);
 
-        if (this._options.aws_appsync_graphqlEndpoint && this._options.aws_appsync_region &&
-            !this._pluggables.find(p => p.getProviderName() === 'AWSAppSyncProvider')) {
-            this.addPluggable(new AWSAppSyncProvider());
```
Comment (powerful23) [inline packages/pubsub/src/PubSub.ts]: Could be a breaking change.

File: packages/core/src/constants.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+/**
+ * This Symbol is used to reference an internal-only PubSub provider that 
+ * is used for AppSync/GraphQL subscriptions in the API category.
+ */
+export const INTERNAL_AWS_APPSYNC_PUBSUB_PROVIDER = Symbol('INTERNAL_AWS_APPSYNC_PUBSUB_PROVIDER');
```
Comment (powerful23) [inline packages/core/src/constants.ts]: Need to expose it to the developers if they want to use `AWSAppsyncProvider`

**Pull request merged**